presto-common/src/main/java/com/facebook/presto/common/type/UserDefinedDistinctType.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common.type;
+
+import com.facebook.presto.common.QualifiedObjectName;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class UserDefinedDistinctType
```

Comment:
hmmm, isn't this just the type signature of `DistinctType`? It's not a type per se. But on the other hand, how likely we can merge this class with `UserDefinedType`? Essentially, it's just to add an optional parent to `UserDefinedType` right?

Reply:
I don't think we should be adding this to UserDefinedType. Mainly because it would diverge from how enums are implemented. For enums, UserDefinedType class is used as an interface when fetching from DB, which stores the typesignature. And we store the actual enum definition in TypeSignature.TypeSignatureParameter using LongEnumMap class.

For distinct type, we reuse UserDefinedType as interface in the DB, and we store UserDefinedDistinctType in TypeSignatureParameter. That's why this has @Json interfaces, to help when we serialize type signatures.

We can probably merge these two classes, by storing raw information rather than a big signature in the userdefinedtype, but that should be done for enums as well. Doing it for partially distinct types and not enums will lead to more confusion. Perhaps done better in a new PR